### ISSUE TITLE
joint_tests() with nested structures; and other points of confusion

### ISSUE THREAD

**Issue by rvlenth** [rvlenth/emmeans#266]
This is somewhat related to Issue #265, but with a twist, so I am posting this as a new issue in case it is of interest to someone.

When we have a nested model, and **emmeans** version <= 1.5.4, we tend to get the same test more than once in the joint_tests listing. To illustrate, here is an utterly silly model for the `auto.noise` data:
```r
> require(emmeans)
> noise.nested <- lm(noise/10 ~ (size / type) * side, data = auto.noise)

> RG <- ref_grid(noise.nested)
NOTE: A nesting structure was detected in the fitted model:
    type %in% size

> joint_tests(RG)
 model term     df1 df2 F.ratio p.value note
 size             2  24 893.190 <.0001      
 type             5  24 382.790 <.0001   d  
 side             1  24   0.048 0.8291      
 size:type        5  24 382.790 <.0001   d  
 size:side        2  24  44.333 <.0001      
 type:side       11  24 184.048 <.0001   d  
 size:type:side  11  24 184.048 <.0001   d  

d: df1 reduced due to linear dependence 
```
Note that the tests for `type` and `size:type` are identical. As are `type:side` and `size:type:side`. That is because of two things:

  1. `joint_tests()` works by looking at all possible combinations of factors. That in turn is because it isn't set up to work for just models, but for any `emmGrid` object.
  2. `type` is nested in `size`; so really, we shouldn't be looking at `type` as a main effect.

Point (1) is what relates to #265. In addressing that, I added code to bypass interactions that aren't in the model (in the case of factors that are model predictors, as opposed to others that come up in creating an `emmGrid` object). But that code does not screen-out *main effects* that are not supposed to be there. That is what is needed to resolve this issue.

This issue can really compound when we use a model to create some other object; for example:

```r
> CON <- pairs(RG, simple = "side")

> str(CON)
'emmGrid' object with variables:
    size = S, M, L
    type = Std, Octel
    contrast = L - R
Nesting structure:  contrast %in% (size*type), type %in% size

> joint_tests(CON)
 model term         df1 df2 F.ratio p.value note
 size                 2  24  44.333 <.0001      
 type                 5  24  22.105 <.0001   d  
 contrast             5  24  22.105 <.0001   d  
 size:type            5  24  22.105 <.0001   d  
 size:contrast        5  24  22.105 <.0001   d  
 type:contrast        5  24  22.105 <.0001   d  
 size:type:contrast   5  24  22.105 <.0001   d  

d: df1 reduced due to linear dependence 
```
Now we have the same test repeated a whole bunch of times. Note, by the way, that in `CON`, the "response" is now the effect of `side` under different combinations of the other factors. So the test of `size` above really is a test of `size:side`, and matches that test in the joint tests for `RG`. The rest is confusing to interpret.

**Comment by rvlenth** [comment]
OK. I added code that looks at the `model.info$nesting` slot. It will suppress computing joint tests for any factor combinations that do not include all nesting factors involved. Now we get (with the objects defined above):
```r
> joint_tests(RG)
 model term     df1 df2 F.ratio p.value note
 size             2  24 893.190 <.0001      
 side             1  24   0.048 0.8291      
 size:type        5  24 382.790 <.0001   d  
 size:side        2  24  44.333 <.0001      
 size:type:side  11  24 184.048 <.0001   d  

d: df1 reduced due to linear dependence 

> joint_tests(CON)
 model term         df1 df2 F.ratio p.value note
 size                 2  24  44.333 <.0001      
 size:type            5  24  22.105 <.0001   d  
 size:type:contrast   5  24  22.105 <.0001   d  

d: df1 reduced due to linear dependence 
```
The results for `RG` are now correct, in that we don't have duplicate results. The results for `CON` are not perfect, but at least less cluttered. That 2nd and 3rd test are still confusing. One might think the test of `size:type` should be the same as `size:type:side`, but 6 d.f. of those 3-way-interaction effects have gone by the wayside. (And what can we expect? `CON` consists of only 6 paired differences).

Anyway, at least we have reasonable, non-duplicated results for the model itself. So that's a start.

**Comment by rvlenth** [comment]
Hmmm, I thought I had posted an update. I now have everything working right. For the objects created earlier, we now have:
```r
> joint_tests(RG)
 model term     df1 df2 F.ratio p.value
 size             2  24 893.190 <.0001 
 side             1  24   0.048 0.8291 
 size:type        3  24  42.524 <.0001 
 size:side        2  24  44.333 <.0001 
 size:type:side   3  24   7.286 0.0012 

> joint_tests(RG, by = "side")
side = L:
 model term df1 df2 F.ratio p.value
 size         2  24 651.714 <.0001 
 size:type    3  24  31.048 <.0001 

side = R:
 model term df1 df2 F.ratio p.value
 size         2  24 285.810 <.0001 
 size:type    3  24  18.762 <.0001 


> joint_tests(CON)
 model term df1 df2 F.ratio p.value
 size         2  24  44.333 <.0001 
 size:type    3  24   7.286 0.0012 
```
Note in the results for `CON`, these tests match those for `size:side` and `size:type:side` in the joint tests for `RG`. This is how it should be.

There had been issues with the code for nested contrasts not supporting the `interaction` specification, as well as some minor bookkeeping matters. See [the corrected version](https://github.com/rvlenth/emmeans/commit/dadda5158cf68f6e0cc597957199a3d4a4e521db) for details.

**Comment by rvlenth** [comment]
Closing. Revised code is included in emmeans 1.5.5-1
